sycl/include/sycl/ext/intel/experimental/esimd/math.hpp

```diff
@@ -206,7 +207,7 @@ rol(T1 src0, T2 src1) {
   using ComputationTy = __ESIMD_DNS::computation_type_t<T1, T2>;
```

Comment:
'computation_type' is good for operations like 'add', it gives 'int64_t' for '(int32_t)a + (int64_t)b'.
This logic is incorrect for 'shift' or 'rotate' operators.

 int8_t val = 5;
 int64_t count = 8; // expected result is 5; 5 just rotates one 8-bit cycle.
 int64_t res = rol(a, count); // expected result is '7'

Using 'computation_type' would cause this:
 int8_t val = 0x5;
 int64_t common_type_val = val;
 int64_t count = 8; // expected result is 5; 5 just rotates one 8-bit cycle.
 int64_t res = rol(common_type_val, count); // result is 0x500, which is wrong.

Reply:
> all the arguments are of the same type. which requires some sort of type promotion

Yes, I agree, the type promotion is needed. It should be a bit different though.
If our esimd::lsr would mimic the regular C++ logical-shift-left, then I see the types are promoted to either int32 (if the 1st arg is byte/short/int) or int64 if the first operand is int64.
  ```
  #include <iostream>
  #include <cstdint>
  
  uint8_t  seven_8 = 7;
  uint64_t seven_64 = 7;
  uint64_t thirty_one_64 = 31;
  
  uint8_t val_8 = 0x0f;
  uint32_t val_32 = 0x0f;
  
  uint64_t result_8_8;
  uint64_t result_8_64;
  uint64_t result_32_64;
  
  int main() {
    result_8_8 = (val_8 << seven_8);
    std::cout << "(uint8_t)0x0F << (uint8_t)7 = 0x" << std::hex << result_8_8 << std::endl;
  
    result_8_64 = (val_8 << seven_64);
    std::cout << "(uint8_t)0x0F << (uint64_t)7 = 0x" << std::hex << result_8_64 << std::endl;
  
    result_32_64 = (val_32 << thirty_one_64);
    std::cout << "(uint32_t)0x0F << (uint64_t)31 = 0x" << std::hex << result_32_64 << std::endl;
    return 0;
  }
```
gives such output. See that the 3rd operation was not promoted to uint64_t:
(uint8_t)0x0F << (uint8_t)7 = 0x780
(uint8_t)0x0F << (uint64_t)7 = 0x780
(uint32_t)0x0F << (uint64_t)31 = 0x80000000